doc/api/n-api.md

```diff
@@ -0,0 +1,3151 @@
+purposes:
+- Get additional information about the context in which the callback was 
+  invoked
+- Set the return value of the callback
+
+#### *napi_callback*
+Function pointer type for user-provided native functions which are to be 
+exposed to JavaScript via N-API. Your callback function should satisfy the 
+following signature:
+```C
+typedef void (*napi_callback)(napi_env, napi_callback_info);
+```
+
+#### *napi_finalize*
+Function pointer type for add-on provided functions that allow the user to be
+notified when externally-owned data is ready to be cleaned up because the
+object with which it was associated with, has been garbage-collected. The user
+must provide a function satisfying the following signature which would get
+called upon the object's collection. Currently, `napi_finalize` can be used for
+finding out when objects that have external data are collected.
+
+
+#### napi_async_execute_callback
+
+Function pointer used with the functions that support asynchronous
+operations. Your callback function should statisfy the following signature:
+
+```C
+typedef void (*napi_async_execute_callback)(napi_env env,
+                                            void* data);
+```
+
+#### napi_async_complete_callback
+
+Function pointer used with the functions that support asynchronous
+operations. Your callback function should statisfy the following signature:
+
+```C
+typedef void (*napi_async_complete_callback)(napi_env env,
+                                             napi_status status,
+                                             void* data);
+```
+
+## Error Handling
+N-API uses both return values and Javascript exceptions for error handling.
+The sections which follow explain the approach for each case.
+
+### Return values
+
+All of the N-API functions share the same error handling pattern.  The
+return type of all API functions is `napi_status`.
+
+The return value will be `napi_ok` if the request was successful and
+no uncaught JavaScript exception was thrown. If an error occurred AND
+an exception was thrown, the `napi_status` value for the error
+will be returned.  If an exception was thrown, and no error occurred,
+`napi_pending_exception` will be returned.
+
+In cases where a return value other than `napi_ok` or
+`napi_pending_exception` is returned, you must call
+`napi_is_exception_pending()` to check if an exception is pending.
+See the section on exceptions for more details.
+
+The full set of possible napi_status values is defined
+in `napi_api_types.h`.
+
+Thr `napi_status` return value provides a vm-independent representation of
+the error which occurred.  In some cases it is useful to be able to get
+more detailed information, including a string representing the error as well as
+vm(engine)-specific information.  **NOTE:**  you should not rely on the
+content or format of any of the extended information as it is
+not subject to SemVer and may change at any time. It is intended
```

Comment:
We do have some control in that the information will be put together by the 'port' of the N-API implementation for each VM (ex V8, Chakra) which is under out control.  However the information that makes sense may vary across VM's and it may incorporate information under the control of the VM.